### ISSUE TITLE
Seperate monolithic Dockerfile

### ISSUE THREAD

**Issue by adrianord** [lspcontainers/dockerfiles#15]
I feel that a monolithic Dockerfile might not fit this project like it would a mono-repo with services that depend on each other or share a common build paths. This couples languages very early on even if their build paths have nothing similar and (in my opinion) hurts readability. This will also cause an increasing amount of merge conflicts as people submit new languages at the same time and as the file grows. This is a very common problem with dotnet's solution file early on in a project when everyone is modifying it to make new projects in the solution. I don't think it does anything to optimize the actual image sizes as they are still larger than when we had first started, at least the images based on `vscode-languageserver-node`. We had also removed the `.env` and node_base images to reduce coupling in the past.

To summarize:
  - Leads to unnecessary coupling which could hurt maintainability and readability
  - Some languages in the Dockerfile share no common build paths
  - Even if they share a common build path like `vscode-languageserver-node` it might still be a good idea not to couple the languages for ease maintainability.
  - A monolithic Dockerfile that everybody must touch will cause merge conflicts for contributors adding their language to it at the same time.

(A secondary concern is this file grows so large with so many languages that the project require Git LFS as I have seen this happen with regularly text file, but this would happen very far down the line.)

**Comment by erkrnt** [comment]
So the reasoning for the monolithic Dockerfile was I was trying out different builds - ways of sharing stages as dependencies and more. My main goal was to try and optimize the images as much as possible breaking everything for the best result and then taking steps back on what I feel we can sacrifice on. 

For example with the `vscode-langserver` implementations:

- The `node:alpine` image adds about 30mb of overhead than the normal `alpine` image
- However, to optimally use the `alpine` image we would have to strongly consider using base images which I don't know if I want to introduce that into the design yet.

My biggest concern hands down with a monolithic Dockerfile is maintainability and essentially the developer experience which is why we will not be keeping this design. However, I was purposely took the crack a few eggs and change the game to see what benefits we could get. I also want to lean heavily into Buildx and it's improvements over the normal builder (things like docker-compose don't use).

I'll open a PR for this change really quickly and get it back to the original implementation with ONE CAVEAT:

For now, I think it's good to not try and overly optimize with base images or stages outside of the LSP Dockerfile. Here is why:

1. You can achieve the same small image with the same lines in each files (though this is duplication)
2. It's more straightforward for people contributing first time etc (which is also why I wanted to try the monolith as a test)

**Comment by adrianord** [comment]
I completely agree, I know I introduced the change to try to make a node_base image, but I think for maintainability sake it's best to have every LSP have it's own full Dockerfile for the time being, especially since we're so early in the process and we want as many people contributing languages to catch up with LspInstall as possible.

I also want to recommend putting the languages into a languages folder just to keep the root of the project a little more tidy.
Ex. `languages/typescript-language-server/Dockerfile`

**Comment by erkrnt** [comment]
> I also want to recommend putting the languages into a languages folder just to keep the root of the project a little more tidy.
> Ex. `languages/typescript-language-server/Dockerfile`

Yeah I agree - we don't want all the languages at the root. Too messy.
